### PR TITLE
Allowance revamp

### PR DESCRIPTION
**Pull request by cristovaoth**
This PR migrates the AllowanceModule to a more modern development environment:
* JavaScript code migrated to Typescript latest
* Contract environment migrated to Hardhat latest
* Using ethers@v6 latest
* Migrated test cases to run on the configured environment
* Code quality tools: prettier, solhint, eslint, solcover
* Uses hardhat-deploy for deterministic deployment of the Allowance Singleton
* I also included typechain, which generates typescript types for contracts, and comes in very handy when writing tests

Some additional points to consider:
* The deployment scripts have been executed on the Goerli network. Can be viewed here [Goerli Deployment](https://goerli.etherscan.io/address/0xd9d7259222c47ee0a7ece7d528e58418462072c0#code).
* I translated the tests 1 to 1, preserving the original test intents, did not increase coverage or rework tests in any way.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
-node_modules/
+node_modules
+build
+coverage
+coverage.json
+deployments
+typechain-types
+*.log
+.env
+.DS_Store
```
Comment (mmv08) [inline .gitignore:9]: nit: .DS_Store is duplicated

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "editor.tabSize": 2,
+  "editor.formatOnSave": true,
+  "cSpell.userWords": [], // only use words from .cspell.json
```
Comment (mmv08) [inline .vscode/settings.json:4]: There are no comments in json innit?

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (mmv08) [inline .vscode/settings.json:1]: i'm not sure if this file should be a part of the repo

File: allowances/test/signature.spec.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        transfer.amount,
+        transfer.paymentToken,
+        transfer.payment,
+        transfer.nonce
+      )
+    ).to.equal(transferHash)
+  })
+})
+
+function calculateTransferHash(
```
Comment (mmv08) [inline allowances/test/signature.spec.ts:62]: You can use https://docs.ethers.org/v6/api/hashing/#TypedDataEncoder in this function to simplify it

File: .gitignore (hunk 1 above)
Comment (cristovaoth) [reply .gitignore]: Crucial

File: .vscode/settings.json (hunk 3 above)
Comment (cristovaoth) [reply .vscode/settings.json]: If you not sure, we aint sure, innit?

Comment (cristovaoth) [reply .vscode/settings.json]: Removed

File: allowances/test/signature.spec.ts (hunk 4 above)
Comment (cristovaoth) [reply allowances/test/signature.spec.ts:43]: Ok, just addressed this. We no longer assert on the separators, only on the hash as a whole.

File: allowances/hardhat.config.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    sources: 'contracts',
+  },
+  solidity: {
+    compilers: [
+      {
+        // for the Allowance contact
+        version: '0.7.6',
+        settings: {
+          optimizer: {
+            enabled: true,
```
Comment (mmv08) [inline allowances/hardhat.config.ts:38]: the optimizer was disabled with truffle

File: allowances/test/allowanceRecurring.spec.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
+import { loadFixture, mine } from '@nomicfoundation/hardhat-network-helpers'
+import { expect } from 'chai'
+
+import execAllowanceTransfer from './test-helpers/execAllowanceTransfer'
+import execSafeTransaction from './test-helpers/execSafeTransaction'
+import setup from './test-helpers/setup'
+
+describe('AllowanceModule allowanceRecurring', async () => {
```
Comment (mmv08) [inline allowances/test/allowanceRecurring.spec.ts:8]: it doesn't seem like the top level function needs to be `async`

File: allowances/test/allowanceSingle.spec.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+import { parseUnits, ZeroAddress } from 'ethers'
+import hre from 'hardhat'
+
+import execAllowanceTransfer from './test-helpers/execAllowanceTransfer'
+import execSafeTransaction from './test-helpers/execSafeTransaction'
+import setup from './test-helpers/setup'
+
+const OneEther = parseUnits('1', 'ether')
+
+describe('AllowanceModule allowanceSingle', async () => {
```
Comment (mmv08) [inline allowances/test/allowanceSingle.spec.ts:12]: it doesn't seem like the top level function needs to be async

**Review by mmv08 — APPROVED** [review]
Looks boss mate

**Pull request merged**